File: Assets/ToLua/Lua/user/ui/chat/controller/chat_controller.lua
local t = {}
local name = 'chat_controller'

require 'chat_pb'

local function Start ()
  netmanager.RegisterProtocol(MSG.ChatResp, t.ChatResp)
  netmanager.RegisterProtocol(MSG.ChatInfoResp, t.ChatInfoResp)
  netmanager.RegisterProtocol(MSG.CheckRoleGuildResp, t.CheckRoleGuildResp)
        
  gamemanager.RegisterCtrl(name, t)
    -- t.ChatInfoReq()

end

function OpenChat()

  if gamemanager.GetModel('chat_model').hasGuild then
    -- print('不在公会')
    netmanager.SendProtocol(MSG.ChatInfoReq)
  else
    -- print('在公会')
    t.CheckRoleGuildReq()
  end
 
  if t.view then
    t.view=dofile('ui/chat/view/chat_view')
    t.view.Open()
    gamemanager.GetCtrl('friend_controller').FriendListReq()

  else
    -- t.ChatInfoReq()
    t.view=dofile('ui/chat/view/chat_view')
    t.view.Open()
  end

  -- t.CheckRoleGuildReq()
end

function t.IsViewOpen()
  if t.view then
    return t.view.open
  end
end
--------------------------protocol---------------------
function t.ChatInfoReq()
  netmanager.SendProtocol(MSG.ChatInfoReq)
end

function t.ChatReq(chatType,revRoleName,content)
  local req = chat_pb.ChatReq()
  req.chatType = chatType
  req.revRoleName = revRoleName
  req.content = content
  -- print('ChatReq:',chatType,revRoleName,content)
  netmanager.SendProtocol(MSG.ChatReq,req)
end

function t.CheckRoleGuildReq()
  netmanager.SendProtocol(MSG.CheckRoleGuildReq)
end

function t.ChatResp()
  local resp = chat_pb.ChatResp()
  resp:ParseFromString(netmanager.GetProtocolData())
  -- print('响应聊天')
  gamemanager.GetModel('chat_model').ChatResp(resp)
end

function t.ChatInfoResp()
  local resp = chat_pb.ChatInfoResp()
  resp:ParseFromString(netmanager.GetProtocolData())
  -- print('响应聊天列表')
  gamemanager.GetModel('chat_model').ChatInfoResp(resp)
end

function t.CheckRoleGuildResp()
  local resp = chat_pb.CheckRoleGuildResp()
  resp:ParseFromString(netmanager.GetProtocolData())
  -- print('响应聊天列表')
  gamemanager.GetModel('chat_model').CheckRoleGuildResp(resp)
end

Start ()
return t

File: Assets/Code/Logic/UI/Chat/Controller/ChatController.cs
using UnityEngine;
using System.Collections;
using Logic.UI.Chat.Model;
using Logic.Protocol.Model;
using Logic.Enums;
using Logic.Game.Model;


namespace Logic.UI.Chat.Controller
{
	public class ChatController : SingletonMono<ChatController> {
		
		void Awake()
		{
			instance = this;
		}
		void Start ()
		{
			Observers.Facade.Instance.RegisterObserver(((int)MSG.ChatResp).ToString(),LOBBY2CLIEN_Chat_RESP_handler);
			Observers.Facade.Instance.RegisterObserver(((int)MSG.ChatInfoResp).ToString(),LOBBY2CLIEN_ChatList_RESP_handler);
		}
		
		void OnDestroy ()
		{
			if (Observers.Facade.Instance != null)
			{
				Observers.Facade.Instance.RemoveObserver(((int)MSG.ChatResp).ToString(),LOBBY2CLIEN_Chat_RESP_handler);
				Observers.Facade.Instance.RemoveObserver(((int)MSG.ChatInfoResp).ToString(),LOBBY2CLIEN_ChatList_RESP_handler);
			}
		}

		#region client to server
		public void CLIENT2LOBBY_ChatList_REQ()
		{
			ChatInfoReq req = new ChatInfoReq();
			Protocol.ProtocolProxy.instance.SendProtocol(req);
		}
		public void CLIENT2LOBBY_SendMessage_REQ(ChatInfo info)
		{
			ChatReq req = new ChatReq();
			req.chatType = (int)info.chatType;
			req.revRoleName = info.talkerName;
			req.content = info.content;
			//ChatProxy.instance.AddChatInfo(info);
			//ChatProxy.instance.UpdateChatByProtocol(info.isMe );
			Protocol.ProtocolProxy.instance.SendProtocol(req);
		}
		#endregion

		#region server to client
		public bool LOBBY2CLIEN_Chat_RESP_handler(Observers.Interfaces.INotification note)
		{
			ChatResp resp = note.Body as ChatResp;
			if(resp.chatType == (int)ChatType.World)
			{
				ChatInfo info = new ChatInfo();
				info.chatType = ChatType.World;
				info.content = resp.content;
				info.talkerName = resp.sendRoleName;
//				info.isMe = GameProxy.instance.PlayerInfo.name.Equals(resp.sendRoleName);
				info.isMe = GameProxy.instance.AccountName.Equals(resp.sendRoleName);
				ChatProxy.instance.AddChatInfo(info);
				ChatProxy.instance.UpdateChatByProtocol(info.isMe );
			}else if(resp.chatType == (int)ChatType.System)
			{
				string s = resp.content;
				if(resp.noticeNo!= 0)
				{
					SystemNoticeProxy.instance.AddSystemNotice(resp.noticeNo,resp.noticParams);
				}else
				{
					SystemNoticeProxy.instance.AddSystemNotice(s);
				}
			}
			return true;
		}
		public bool LOBBY2CLIEN_ChatList_RESP_handler(Observers.Interfaces.INotification note)
		{
			ChatInfoResp resp = note.Body as ChatInfoResp;

			ChatProxy.instance.AddChatInfoList(resp.chatList);
			return true;
		}
		#endregion
	}
}

